### PR TITLE
METRON-1945: Metron MPack support for Knox SSO setup

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This PR is based on https://github.com/apache/metron/pull/1275 and adds Ambari MPack support for enabling Metron on Knox.  This makes the setup process simpler, consisting of only a couple steps.

### Changes Included

- Added code to the Ambari METRON_CLIENT component that performs the necessary setup/installation for Metron on Knox
- Added `metron.knox.enabled` and `metron.knox.sso.pubkey` properties to Ambari
- Added Metron Knox topology files as templates in Ambari so the correct hosts can be configured automatically
- Added `app-config.json` files as templates in Ambari so the correct paths can be configured automatically
- Added a check for LDAP, throwing an error message if it is not enabled
- Added logic to populate the Spring active profiles and options on REST startup

### Testing

This has been tested in full dev.  The setup steps are similar to https://github.com/apache/metron/pull/1275 but should be substantially simpler with less manual work involved:

1. Add Knox as a Service in Ambari (Admin > Stacks and Versions > Add Service next to Knox)
2. Verify the LDAP check is working.  Navigate to `Ambari > Metron > Configs > Security` and set `Knox Enabled` to true.  Restart the Metron REST component and it should fail with this error:
```
Enabling Metron with Knox requires LDAP authentication.  Please set `LDAP Enabled` to true in the Metron Security tab.
```
Change `Knox Enabled` back to false and save.

3. Follow the instructions [here](https://github.com/apache/metron/tree/master/metron-deployment/development#knox-demo-ldap) to enable the Knox LDAP in full dev.  You should be able to authentication with REST using guest/guest-password after this is done.
4. A Metron Client instance is required on all Knox Gateway hosts.  Navigate to `Ambari > Hosts > node1` and select the `Install clients` option in the dropdown next to the clients, at the bottom of the `Components` section.  This should install the Metron Client.
5. Get the Knox public key with the following command on full dev:
```
openssl s_client -connect node1:8443 < /dev/null | openssl x509 | grep -v 'CERTIFICATE' | paste -sd "" -
```
Copy the output of that command and paste to the Ambari setting at `Metron > Configs > Security > Knox SSO Public Key`.  Change the `Knox Enabled` setting right above that to true and save.

6. You should be prompted to restart Metron Client, Metron REST, Metron Alerts UI and Metron Management UI.  Follow the prompt and restart these components.

After REST comes back up Metron should be enabled for Knox.  You should be able to follow the instructions in the https://github.com/apache/metron/pull/1275 description for verifying that SSO is working for all services.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py [hunk 1]
```diff
@@ -51,23 +51,24 @@
...
 # Construct the profiles as a temp variable first. Only the first time it's set will carry through
-metron_spring_profiles_temp = config['configurations']['metron-rest-env']['metron_spring_profiles_active']
-if config['configurations']['metron-security-env']['metron.ldap.enabled']:
-    if metron_spring_profiles_temp:
-        metron_spring_profiles_active = metron_spring_profiles_temp + ',ldap'
+metron_spring_profiles_active = config['configurations']['metron-rest-env']['metron_spring_profiles_active']
+metron_ldap_enabled = config['configurations']['metron-security-env']['metron.ldap.enabled']
+if metron_ldap_enabled:
+    if not len(metron_spring_profiles_active) == 0:
+        metron_spring_profiles_active += ',ldap'
```
[the remark this answers is not in the document]
Comment (anandsubbu) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Setting it here will avoid the manual step of adding `knox` to the active spring profiles and restart REST services again.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json [hunk 2]
```diff
@@ -1687,6 +1714,18 @@
...
+      {
+        "config": "metron-security-env/metron.knox.enabled",
+        "widget": {
+          "type": "toggle"
+        }
+      },
+      {
+        "config": "metron-security-env/metron.knox.sso.pubkey",
+        "widget": {
+          "type": "text-field"
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json]: Done.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Are you sure you had to manually add `knox` to the active spring profiles?  This should happen automatically on line 465 of this file.  If you did have to add it, then there is a bug somewhere.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/metron_service.py [hunk 3]
```diff
@@ -588,3 +589,35 @@ def check_indexer_parameters():
...
+            unset KNOX_GROUP; \
+            unset KNOX_HOME;"""
+        cmd = template.format(params.knox_home, params.knox_user, params.knox_group, params.metron_home)
+
+        Execute(cmd)
+
+def is_metron_knox_installed(params):
+    return os.path.isfile(params.metron_knox_installed_flag_file)
+
+def set_metron_knox_installed(params):
```
[the remark this answers is not in the document]
Comment (anandsubbu) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/metron_service.py]: I checked this behavior on a kerberized environment. Here is the folder structure on a node _without_ Metron services:

```
[root@metron-1 ~]# cd /usr/metron/0.7.0/
[root@metron-1 0.7.0]# ll
total 4
-rwxr-xr-x. 1 metron metron 688 Dec 21 10:18 client_jaas.conf
```
Restarting Metron Client service is still failing with this error:
```
    raise Fail("Applying %s failed, parent directory %s doesn't exist" % (self.resource, dirname))
resource_management.core.exceptions.Fail: Applying File['/usr/metron/0.7.0/config/zookeeper/../metron_knox_installed'] failed, parent directory /usr/metron/0.7.0/config/zookeeper/.. doesn't exist

```

The `config/zookeeper` within `$METRON_HOME` needs to be created before we write the flag file.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anandsubbu) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Yup.. just validated that we add `knox` to the active spring profiles. Cheers! 

Here's the relevant snippet from restarting REST:

```
<snip>
2018-12-21 12:59:36,411 - Execute['export METRON_JDBC_PASSWORD=[PROTECTED];export JAVA_HOME=/usr/jdk64/jdk1.8.0_112;export METRON_REST_CLASSPATH=;export METRON_INDEX_CP=;export METRON_LOG_DIR=/var/log/metron;export METRON_PID_FILE=/var/run/metron/metron-rest.pid;export HDP_VERSION=2.6.5.1050-37;export METRON_RA_INDEXING_WRITER=Elasticsearch;export METRON_LDAP_PASSWORD=[PROTECTED];export METRON_LDAP_SSL_TRUSTSTORE_PASSWORD=[PROTECTED];/usr/metron/0.7.0/bin/metron-rest.sh;unset METRON_JDBC_PASSWORD;unset METRON_LDAP_PASSWORD;unset METRON_LDAP_SSL_TRUSTSTORE_PASSWORD;'] {'logoutput': True, 'not_if': 'ls /var/run/metron/metron-rest.pid >/dev/null 2>&1 && ps -p  >/dev/null 2>&1', 'user': 'metron', 'timeout': 60}
METRON_VERSION=0.7.0
METRON_HOME=/usr/metron/0.7.0
METRON_SYSCONFIG=/etc/default/metron
METRON_SYSCONFIG=/etc/default/metron
Default metron-rest jar is: /usr/metron/0.7.0/lib/metron-rest-0.7.0.jar
METRON_SPRING_PROFILES_ACTIVE=dev,ldap,knox
METRON_CONFIG_LOCATION= --spring.config.location=/usr/metron/0.7.0/config/rest_application.yml,classpath:/application.yml
METRON_PROFILES_ACTIVE= --spring.profiles.active=dev,ldap,knox
Metron indexing jar is: /usr/metron/0.7.0/lib/metron-elasticsearch-0.7.0-uber.jar
<snip>
```

**Comment by anandsubbu** [comment]
When I click in Ambari -> Metron service -> Quick Links -> any of the links, it does not get redirected to the Knox gateway. Does this need to be fixed as well?

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/metron_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/metron_service.py]: Thanks @anandsubbu.  The latest commit creates this directory if it doesn't already exist.

**Comment by merrimanr** [comment]
I also pushed a commit out that exposes the Knox SSO Token TTL as an Ambari setting and changes the default to 5 minutes.  The default was 30 seconds and I can't imagine anyone would ever leave it set to that.  Is 5 minutes still too short or should it be expected that an administrator will need to change it?  Any thoughts on what this should be?

I tested this again in full dev and the TTL setting works as expected.

**Comment by merrimanr** [comment]
@anandsubbu as far as I know Quick Links must be changed manually.  There is a section in the README explaining this.

I believe all feedback has been addressed.  Let me know if I missed something.

**Comment by merrimanr** [comment]
The only example I can find is the SSO topology that ships with Knox and it's set to 30 seconds.

**Comment by merrimanr** [comment]
You're correct.  Fortunately it's only the testing instructions that were incorrect.  Fixed.

**Comment by mmiklavc** [comment]
@merrimanr Great job on a big submission. I ran this up in full dev and everything seems to be working properly, including timeouts and logouts. I also see @anandsubbu has put this through its paces in a multi-node deployment, which is great. +1 from me pending the latest Travis build succeeding. You also have a +1 from on https://github.com/apache/metron/pull/1275, so this feature should be ready to roll.

**Comment by anandsubbu** [comment]
+1 
Looks good @merrimanr !

**Pull request closed without merging**